contrib/epee/include/net/abstract_tcp_server2.inl

```diff
@@ -790,66 +771,108 @@ PRAGMA_WARNING_DISABLE_VS(4355)
       if(ec == boost::asio::error::operation_aborted)
         return;
       MDEBUG(context << "connection timeout, closing");
-      self->close();
+      self->shutdown(false); // already timed out waiting for send, bail on flush
     });
   }
   //---------------------------------------------------------------------------------
   template<class t_protocol_handler>
-  bool connection<t_protocol_handler>::shutdown()
+  void connection<t_protocol_handler>::finish_shutdown()
   {
-    CRITICAL_REGION_BEGIN(m_shutdown_lock);
-    if (m_was_shutdown)
-      return true;
-    m_was_shutdown = true;
-    // Initiate graceful connection closure.
-    m_timer.cancel();
-    boost::system::error_code ignored_ec;
-    if (m_ssl_support == epee::net_utils::ssl_support_t::e_ssl_support_enabled)
+    //! non-blocking (ssl?) shutdown resumable coroutine
+    class shutdown_routine : public boost::asio::coroutine
     {
-      const shared_state &state = static_cast<const shared_state&>(get_state());
-      if (!state.stop_signal_sent)
-        socket_.shutdown(ignored_ec);
-    }
+      boost::shared_ptr<connection> self;
+
+    public:
+      explicit shutdown_routine(boost::shared_ptr<connection> self)
+        : boost::asio::coroutine(), self(std::move(self))
+      {}
+
+      shutdown_routine(const shutdown_routine&) = default;
+      shutdown_routine(shutdown_routine&&) = default;
+      ~shutdown_routine() = default;
+
+      void operator()(const boost::system::error_code = {})
+      {
+        // A "forced" (unclean) shutdown can be called while waiting
+        if (!self || status::cleanup <= self->m_shutdown_status)
+          return;
+
+        const boost::shared_ptr<connection>& self_copy = self; // for lambda
```

Comment:
Why not a copy then ?

Reply:
The compiler is required to do another atomic inc/dec if this line is by copy - it can't be optimized out. Overkill as usual, but achieves the same goal easily.

Reply:
It is after all just a const reference (an alias). If the pointer type was to be changed, then the copy could fail even.